programs/pluto/kernel_xfrm.c

```diff
@@ -2536,6 +2563,62 @@ static err_t xfrm_migrate_ipsec_sa_is_enabled(struct logger *logger)
 	}
 }
 
+static bool netlink_poke_ipsec_offload_policy_hole(struct nic_offload *nic_offload, struct logger *logger)
+{
+	struct {
+		struct nlmsghdr n;
+		struct {
+			struct xfrm_userpolicy_info p;
+			struct rtattr attr;
+			struct xfrm_user_offload uo;
+		} __attribute__((packed, aligned(4))) nlmsg;
+	} req = {
+		.n = {
+			.nlmsg_flags = NLM_F_REQUEST | NLM_F_ACK,
+			.nlmsg_type = XFRM_MSG_NEWPOLICY,
+			.nlmsg_len = NLMSG_ALIGN(NLMSG_LENGTH(sizeof(req.nlmsg))),
+		},
+		.nlmsg = {
+			.p = {
+				.priority = 1, /* give admin prio 0 as override */
+				.action = XFRM_POLICY_ALLOW,
+				.share = XFRM_SHARE_ANY,
+
+				.lft.soft_byte_limit = XFRM_INF,
+				.lft.soft_packet_limit = XFRM_INF,
+				.lft.hard_byte_limit = XFRM_INF,
+				.lft.hard_packet_limit = XFRM_INF,
+
+				.sel.proto = IPPROTO_UDP,
+				.sel.family = AF_INET,
+				.sel.sport = htons(IKE_UDP_PORT),
+				.sel.dport = htons(IKE_UDP_PORT),
```

Comment:
what about the NAT port 4500 ? Or is UDP Encap not supported?

Reply:
Right i was aware off, UDP Encap is not supported, while at it, shouldn't IKE_UDP_PORT be configurable ? although it is a convention to use 500, however it is not a standard demand to use this port for ike , right ? can you make it configurable if so ?

Reply:
To follow up Paul's point, shouldn't the ip_port be a parameter so that the code opening sockets can poke multiple holes?  Called from bind_iface_endpoint()?
Bare ESP can be negotiated using any port.  The only got-ya is that the UDP IKE packets sent on ports other than 500 always contain the ESP=0 prefix.

Reply:
IKE UDP packets sent to port 500 don't include the four zero bytes at the very start of the message body, but IKE UDP packets to other ports do.  This only matters when IPsec is being encapsulated within UDP.

Given this card doesn't support UDP encap I'm not sure the zero prefix will matter.  Instead all that's needed is for holes to be punched for all relevant UDP ports.

It's described here https://www.rfc-editor.org/rfc/rfc3948.html

Reply:
thanks for the reference will look at it, i think i saw this kind of check in XFRM stack also in linux, true IPSEC UDP encap still not supported for offload